packages/utilities/src/warn.ts

```diff
@@ -70,6 +70,95 @@ export function warnConditionallyRequiredProps<P>(
+  addlOnChangeProp?: keyof P,
+  readOnlyProp?: keyof P
+): void {
+  const hasOnChange = !!(props[onChangeProp] || (addlOnChangeProp && props[addlOnChangeProp]));
+  const isReadOnly = !!(readOnlyProp && props[readOnlyProp]);
+  // != null is the same check used internally by React
+  // https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOMInput.js
+  const hasValue = props[valueProp] != null; // tslint:disable-line:triple-equals
+  const hasDefaultValue = props[defaultValueProp] != null; // tslint:disable-line:triple-equals
+
+  if (hasValue && !(hasOnChange || isReadOnly) && !warningsMap.valueOnChange[componentId]) {
+    warningsMap.valueOnChange[componentId] = true;
+    warn(
+      `Warning: You provided a '${valueProp}' prop to a ${componentName} without an '${onChangeProp}' handler. ` +
+        `This will render a read-only field. If the field should be mutable use '${defaultValueProp}'. ` +
+        `Otherwise, set '${onChangeProp}'${readOnlyProp ? ` or '${readOnlyProp}'` : ''}.`
+    );
+  }
+
+  if (hasValue && hasDefaultValue && !warningsMap.valueDefaultValue[componentId]) {
+    warningsMap.valueDefaultValue[componentId] = true;
+    warn(
+      `Warning: You provided both '${valueProp}' and '${defaultValueProp}' to a ${componentName}. ` +
+        `Form fields must be either controlled or uncontrolled (specify either the '${valueProp}' prop, ` +
+        `or the '${defaultValueProp}' prop, but not both). Decide between using a controlled or uncontrolled ` +
+        `${componentName} and remove one of these props. More info: https://fb.me/react-controlled-components`
+    );
+  }
+}
+
+/**
+ * Warn if a form component is switching between controlled and uncontrolled usage.
+ * The message mimics the warning React gives for this error condition on input elements.
+ * @param componentId - The ID of the form component. Used to prevent showing warnings repeatedly.
+ * @param componentName - The name of the form component.
+ * @param wasControlled - true if the component was originally controlled, false if it was
+ * uncontrolled. Used to determine the error message variant displayed.
+ */
+export function warnControlledUncontrolledSwitch(componentId: string, componentName: string, wasControlled: boolean): void {
+  const oldType = wasControlled ? 'a controlled' : 'an uncontrolled';
+  const newType = wasControlled ? 'uncontrolled' : 'controlled';
+  const warnMap = wasControlled ? warningsMap.controlledToUncontrolled : warningsMap.uncontrolledToControlled;
+  if (!warnMap[componentId]) {
+    warnMap[componentId] = true;
+    warn(
+      `Warning: A component is changing ${oldType} ${componentName} to be ${newType}. ` +
+        `${componentName}s should not switch from controlled to uncontrolled (or vice versa). ` +
+        `Decide between using controlled or uncontrolled for the lifetime of the component. ` +
+        `More info: https://fb.me/react-controlled-components`
+    );
```

Comment:
`warn` will break existing unit tests if it is triggered, and I have a feeling there's a high potential here for controlled vs. uncontrolled misuse as the component itself was encouraging it. Is `warn` the proper level of action here? Will we communicate the potential for breaking change with the merge of this PR?

Reply:
I fixed the tests that broke due to using the component incorrectly. :P You're right that there would definitely be some extra communication required when this is merged.